### DISCUSSION TITLE
ImportError: cannot import name 'get_default_executor' from 'airflow.executors'

### DISCUSSION THREAD

**Discussion by user17904884** [apache/airflow#20815] [Q&A]
In Airflow 1.10.15 using celery executor in docker, the "apache-airflow-upgrade-check" package was used for the checks required to upgrade to Airflow 2.2.3. The below image is the output after executing the command. The Airflow Config file is updated as required for Airflow 2.2.3.
<img width="1712" alt="Screenshot 2022-01-11 at 5 27 05 PM" src="https://user-images.githubusercontent.com/97545449/148990251-05290802-6805-487c-afce-7ac6483afa66.png">

The below command was used to upgrade to Airflow 2.2.3 in Airflow Webserver container.
`pip install "apache-airflow[celery]==2.2.3" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.2.3/constraints-3.6.txt" 
`

After upgrading and restarting the docker engine, I am facing the below error while accessing the Airflow Webserver, Scheduler and worker containers.
`ImportError: cannot import name 'get_default_executor' from 'airflow.executors' (/usr/local/airflow/.local/lib/python3.7/site-packages/airflow/executors/__init__.py)
`
<img width="1171" alt="Screenshot 2022-01-11 at 5 48 23 PM" src="https://user-images.githubusercontent.com/97545449/148990523-ff282c96-11fc-42b7-a065-3661d938de9a.png">

The file **/usr/local/airflow/.local/lib/python3.7/site-packages/airflow/executors/init.py** is empty in Airflow 2.2.3 but not in Airflow 1.10.15.

Any advice would be appreciated.

**Comment by potiuk** [comment]
1) Use 3.7 constraints (not 3.6)
2) I think you have two airflow installed in the image - 1.10 and 2 (but in different locations that's why they are clashing). I am not sure how you created/upgraded stuff but generally when you use docker image you should not manually upgrade airflow in the image but either use or build a new image. Seems you tried to PIP-install airflow in the existing 1.10 image which is likely a bad idea.

**Reply by user17904884** [reply]
Thank you @potiuk for the feedback.

I followed the below steps to upgrade to Airflow 2.2.3:

1. I Used [docker-compose file](https://github.com/puckel/docker-airflow/blob/master/docker-compose-CeleryExecutor.yml) to create Airflow 1.10.9 container.
2. Ran a couple of DAGs.
3. I Used `pip install apache-airflow==1.10.15` command to upgrade to 1.10.15 in webserver container. The command successfully installed Airflow 1.10.15. (Python version is 3.7.6)
4. Executed `airflow db upgrade` command to update the metadata database to the latest version.
5. Installed apache-airflow-upgrade-check package and ran `airflow upgrade_check` command. (Changed **rbac** value to True, created a user,  installed required Backport Providers)
6. Then executed `pip install "apache-airflow[celery]==2.2.3" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.2.3/constraints-3.6.txt"` command which installed Airflow 2.2.3. But is resulting in the Import Error related to cannot import name 'get_default_executor'.
    
    Before upgrading I took the backup of Airflow 1.10.15 files and compared the file **usr/local/airflow/.local/lib/python3.7/site-     packages/airflow/executors/init.py** (which is not empty in Airflow 1.10.15) with the file in Airflow 2.2.3 which is empty.

**Reply by potiuk** [reply]
This is wrong. You should simply switch to a different airlfow image in the docker compose, 

Do not upgrade airflow inside running container. Upgrading  anytihin inside running container makes no sense whatsoever.  This is not how docker compose (and docker) works.

If you want to use docker compose, you need to understand what you are doing. We warned about it here: 

https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html#production-readiness

> DO NOT expect the Docker Compose below will be enough to run production-ready Docker Compose Airflow installation using it. This is truly quick-start docker-compose for you to get Airflow up and running locally and get your hands dirty with Airflow. Configuring a Docker-Compose installation that is ready for production requires an intrinsic knowledge of Docker Compose, a lot of customization and possibly even writing the Docker Compose file that will suit your needs from the scratch. It’s probably OK if you want to run Docker Compose-based deployment, but short of becoming a Docker Compose expert, it’s highly unlikely you will get robust deployment with it.

> If you want to get an easy to configure Docker-based deployment that Airflow Community develops, supports and can provide support with deployment, you should consider using Kubernetes and deploying Airflow using Official Airflow Community Helm Chart.

If you want to manually install airflow using PIP, use https://airflow.apache.org/docs/apache-airflow/stable/start/local.html

**Reply by user17904884** [reply]
@potiuk Thank you for the solution and the information.

**Reply by user17904884** [reply]
May I know on how to migrate metadata from one version to another if I change airflow image version in the docker compose file.
Can I just change the airflow image version in docker compose file or should I create a custom image using docker file with some commands for migration and use it?

**Reply by potiuk** [reply]
Docker image does not contain any metadata. It uses the DB whih is outside it. If you had custom image before - you should also customize it now - depends how you created it.
